### ISSUE TITLE
Share pid namespace between containers

### ISSUE THREAD

**Issue by hqhq** [docker/libcontainer#604]
As my approach in https://github.com/docker/docker/pull/13453, got a problem that libcontainer are not support setns to pid namespace properly, if we just setns to pid namespace, the process called setns is still outside of pid namespace.

For starting container, Docker use `nsexec.c` to work around this, but this problem still exists in `docker exec` for now, post this and hope we can discuss a solution for this.

**Comment by rhatdan** [comment]
As you know, when joining a pid namespace, you are required to fork/exec to create a new process within the namespace.

**Comment by hqhq** [comment]
Yeah, that's why I think it's kind of tough job to let a container join an exist pid namespace of another container.
I'm thinking of expand `nsexec.c` that in some scenario we can join some specific ns, like pid ns only. WDYT?

**Comment by rhatdan** [comment]
Sure makes sense to me.

**Comment by avagin** [comment]
I think it's a good idea to expand nsexec.c.
If it isn't a secrete, could you describe a use-case when you need two containers which share pidns?

**Comment by LK4D4** [comment]
Also can be useful for `USERNS`.

**Comment by mrunalp** [comment]
@avagin Having a debug container that can ptrace the process in another container is a good use case.

**Comment by cyphar** [comment]
@mrunalp Surely you'd use `nsenter` in that case?

**Comment by rhatdan** [comment]
@avagin I think people will come up with cases where containers need to see processes inside of other containers.  And maybe nsenter is the best way to do this.  We use SPC containers on ProjectAtomic, which can enter any container.

**Comment by dqminh** [comment]
Right, the problem with nsenter approach is that many times the container doesn't contain necessary debugging tools. IIRC k8 folks want containers inside a pod to be able to signal each other processes ( hence same pid namespace ).

**Comment by syeduguri** [comment]
+1
